Linux kernel version 5.10.y changed the interface for access to kernel space from user programs. The __put_user_x macro is no longer present and needs to be replaced. 

/home/iago/staging/driver/audio_driver.c: In function ‘mr_alsa_audio_pcm_capture_copy_internal’:
/home/iago/staging/driver/audio_driver.c:1210:9: error: implicit declaration of function ‘__put_user_x’; did you mean ‘__put_user_8’? [-Werror=implicit-function-declaration]
 1210 |         __put_user_x(1, val, (unsigned long __user *)src, ret_pu);
      |         ^~~~~~~~~~~~
      |         __put_user_8

--- a/driver/MTConvert.c
+++ b/driver/MTConvert.c
@@ -1320,10 +1320,10 @@
                 {
                     const uint8_t* in = (uint8_t*)input_buffer[ch] + in_pos;
                     #if defined(MTAL_LINUX) && defined(MTAL_KERNEL)
-                        __put_user_x(1, in[1], (unsigned long __user *)out, ret_pu);
+                        ret_pu = __put_user(in[1], out);
                         ret |= ret_pu;
                         out++;
-                        __put_user_x(1, in[0], (unsigned long __user *)out, ret_pu);
+                        ret_pu = __put_user(in[0], out);
                         ret |= ret_pu;
                         out++;
                     #else
@@ -1360,10 +1360,10 @@
                 {
                     const uint8_t* in = (uint8_t*)input_buffer[ch] + in_pos;
                     #if defined(MTAL_LINUX) && defined(MTAL_KERNEL)
-                        __put_user_x(1, in[2], (unsigned long __user *)out, ret_pu);
+                        ret_pu = __put_user(in[2], out);
                         ret |= ret_pu;
                         out++;
-                        __put_user_x(1, in[3], (unsigned long __user *)out, ret_pu);
+                        ret_pu = __put_user(in[3], out);
                         ret |= ret_pu;
                         out++;
                     #else
@@ -1415,13 +1415,13 @@
                 {
                     const uint8_t* in = (uint8_t*)input_buffer[ch] + in_pos;
                     #if defined(MTAL_LINUX) && defined(MTAL_KERNEL)
-                        __put_user_x(1, in[2], (unsigned long __user *)out, ret_pu);
+                        ret_pu = __put_user(in[2], out);
                         ret |= ret_pu;
                         out++;
-                        __put_user_x(1, in[1], (unsigned long __user *)out, ret_pu);
+                        ret_pu = __put_user(in[1], out);
                         ret |= ret_pu;
                         out++;
-                        __put_user_x(1, in[0], (unsigned long __user *)out, ret_pu);
+                        ret_pu = __put_user(in[0], out);
                         ret |= ret_pu;
                         out++;
                     #else
@@ -1461,13 +1461,13 @@
                 {
                     const uint8_t* in = (uint8_t*)input_buffer[ch] + in_pos;
                     #if defined(MTAL_LINUX) && defined(MTAL_KERNEL)
-                        __put_user_x(1, in[1], (unsigned long __user *)out, ret_pu);
+                        ret_pu = __put_user(in[1], out);
                         ret |= ret_pu;
                         out++;
-                        __put_user_x(1, in[2], (unsigned long __user *)out, ret_pu);
+                        ret_pu = __put_user(in[2], out);
                         ret |= ret_pu;
                         out++;
-                        __put_user_x(1, in[3], (unsigned long __user *)out, ret_pu);
+                        ret_pu = __put_user(in[3], out);
                         ret |= ret_pu;
                         out++;
                     #else
@@ -1522,16 +1522,16 @@
                     const uint8_t* in = (uint8_t*)input_buffer[ch] + in_pos;
                     #if defined(MTAL_LINUX) && defined(MTAL_KERNEL)
                         char zero = 0x00;
-                        __put_user_x(1, in[2], (unsigned long __user *)out, ret_pu);
+                        ret_pu = __put_user(in[2], out);
                         ret |= ret_pu;
                         out++;
-                        __put_user_x(1, in[1], (unsigned long __user *)out, ret_pu);
+                        ret_pu = __put_user(in[1], out);
                         ret |= ret_pu;
                         out++;
-                        __put_user_x(1, in[0], (unsigned long __user *)out, ret_pu);
+                        ret_pu = __put_user(in[0], out);
                         ret |= ret_pu;
                         out++;
-                        __put_user_x(1, zero, (unsigned long __user *)out, ret_pu);
+                        ret_pu = __put_user(zero, out);
                         ret |= ret_pu;
                         out++;
                     #else
@@ -1574,16 +1574,16 @@
                     const uint8_t* in = (uint8_t*)input_buffer[ch] + in_pos;
                     #if defined(MTAL_LINUX) && defined(MTAL_KERNEL)
                         char zero = 0x00;
-                        __put_user_x(1, in[1], (unsigned long __user *)out, ret_pu);
+                        ret_pu = __put_user(in[1], out);
                         ret |= ret_pu;
                         out++;
-                        __put_user_x(1, in[2], (unsigned long __user *)out, ret_pu);
+                        ret_pu = __put_user(in[2], out);
                         ret |= ret_pu;
                         out++;
-                        __put_user_x(1, in[3], (unsigned long __user *)out, ret_pu);
+                        ret_pu = __put_user(in[3], out);
                         ret |= ret_pu;
                         out++;
-                        __put_user_x(1, zero, (unsigned long __user *)out, ret_pu);
+                        ret_pu = __put_user(zero, out);
                         ret |= ret_pu;
                         out++;
                     #else
@@ -1606,16 +1606,16 @@
                     const uint8_t* in = (uint8_t*)input_buffer[ch] + in_pos;
                     #if defined(MTAL_LINUX) && defined(MTAL_KERNEL)
                         char zero = 0x00;
-                        __put_user_x(1, in[1], (unsigned long __user *)out, ret_pu);
+                        ret_pu = __put_user(in[1], out);
                         ret |= ret_pu;
                         out++;
-                        __put_user_x(1, in[2], (unsigned long __user *)out, ret_pu);
+                        ret_pu = __put_user(in[2], out);
                         ret |= ret_pu;
                         out++;
-                        __put_user_x(1, in[3], (unsigned long __user *)out, ret_pu);
+                        ret_pu = __put_user(in[3], out);
                         ret |= ret_pu;
                         out++;
-                        __put_user_x(1, zero, (unsigned long __user *)out, ret_pu);
+                        ret_pu = __put_user(zero, out);
                         ret |= ret_pu;
                         out++;
                     #else
@@ -1656,16 +1656,16 @@
                     {
                         const uint8_t* in = (uint8_t*)input_buffer[ch] + in_pos;
                         #if defined(MTAL_LINUX) && defined(MTAL_KERNEL)
-                            __put_user_x(1, in[3], (unsigned long __user *)out, ret_pu);
+                            ret_pu = __put_user(in[3], out);
                             ret |= ret_pu;
                             out++;
-                            __put_user_x(1, in[2], (unsigned long __user *)out, ret_pu);
+                            ret_pu = __put_user(in[2], out);
                             ret |= ret_pu;
                             out++;
-                            __put_user_x(1, in[1], (unsigned long __user *)out, ret_pu);
+                            ret_pu = __put_user(in[1], out);
                             ret |= ret_pu;
                             out++;
-                            __put_user_x(1, in[0], (unsigned long __user *)out, ret_pu);
+                            ret_pu = __put_user(in[0], out);
                             ret |= ret_pu;
                             out++;
                         #else
@@ -1707,16 +1707,16 @@
                     {
                         const uint8_t* in = (uint8_t*)input_buffer[ch] + in_pos;
                         #if defined(MTAL_LINUX) && defined(MTAL_KERNEL)
-                            __put_user_x(1, in[0], (unsigned long __user *)out, ret_pu);
+                            ret_pu = __put_user(in[0], out);
                             ret |= ret_pu;
                             out++;
-                            __put_user_x(1, in[1], (unsigned long __user *)out, ret_pu);
+                            ret_pu = __put_user(in[1], out);
                             ret |= ret_pu;
                             out++;
-                            __put_user_x(1, in[2], (unsigned long __user *)out, ret_pu);
+                            ret_pu = __put_user(in[2], out);
                             ret |= ret_pu;
                             out++;
-                            __put_user_x(1, in[3], (unsigned long __user *)out, ret_pu);
+                            ret_pu = __put_user(in[3], out);
                             ret |= ret_pu;
                             out++;
                         #else
--- a/driver/audio_driver.c
+++ b/driver/audio_driver.c
@@ -1207,9 +1207,7 @@
     {
         int ret_pu;
         char val = 0xf1;
-        __put_user_x(1, val, (unsigned long __user *)src, ret_pu);
-        ret_pu = put_user(val, (unsigned long __user *)src);
-        //put_user(val, (unsigned long __user *)src);
+        ret_pu = __put_user(val, (char *)src);
         switch(nb_logical_bits)
         {
             case 16:
